commit bdd0c625236bc25799bd6f81dcf5d774928e8cb0 (refs/changes/30/1330/3)
Author: Robert Mustacchi <rm@joyent.com>
Date:   2017-01-24T22:55:58+00:00 (2 years, 8 months ago)
    
    MANTA-3112 nginx suffers from port_getn() ETIME bug
    Reviewed by: Patrick Mooney <patrick.mooney@joyent.com>
    Approved by: Patrick Mooney <patrick.mooney@joyent.com>

diff --git a/src/event/modules/ngx_eventport_module.c b/src/event/modules/ngx_eventport_module.c
index dafa27f6..ed6689e4 100644
--- a/src/event/modules/ngx_eventport_module.c
+++ b/src/event/modules/ngx_eventport_module.c
@@ -468,6 +468,16 @@ ngx_eventport_process_events(ngx_cycle_t *cycle, ngx_msec_t timer,
         ngx_time_update();
     }
 
+    /*
+     * There's a long standing condition with event ports that port_getn() may
+     * return ETIME even when events are available. This would happen if we have
+     * specified a timeout to port_getn() without reaching the number of
+     * requested events.
+     */
+    if (n == -1 && err == ETIME && events > 0) {
+	    n = 0;
+    }
+
     if (n == -1) {
         if (err == ETIME) {
             if (timer != NGX_TIMER_INFINITE) {
